common/src/com/thoughtworks/go/config/materials/git/GitMaterial.java

```diff
@@ -106,28 +127,32 @@ public MaterialConfig config() {
     }
 
     public List<Modification> modificationsSince(File baseDir, Revision revision, final SubprocessExecutionContext execCtx) {
+        unshallowIfNeeded(ProcessOutputStreamConsumer.inMemoryConsumer(), revision, baseDir);
         GitCommand gitCommand = getGit(baseDir);
         return gitCommand.modificationsSince(revision);
     }
 
     public MaterialInstance createMaterialInstance() {
-        return new GitMaterialInstance(url.forCommandline(), branch, submoduleFolder, UUID.randomUUID().toString());
+        return new GitMaterialInstance(url.forCommandline(), branch, submoduleFolder, UUID.randomUUID().toString(), shallowClone);
     }
 
     @Override
     protected void appendCriteria(Map<String, Object> parameters) {
         parameters.put(ScmMaterialConfig.URL, url.forCommandline());
         parameters.put("branch", branch);
+        parameters.put("shallowClone", shallowClone);
```

Comment:
> dude, don't you need sleep ? :-D

I'm closer to you than you think ... :)

> Maybe I am wrong, but submoduleFolder is not included in the identity is problematic ...
> How do we guarantee p1 build get submoduleFolder "foo" and p2 build get submoduleFolder "bar"?

Could be. My point was submoduleFolder doesn't seem to be something you can set in config. I couldn't find it in the [schema](https://github.com/gocd/gocd/blob/eb0ae575fbb0a30f59ed646cd45d16eb57d26aa5/config/config-server/resources/cruise-config.xsd). So, I don't know what it is for, really. Do you?